### DISCUSSION TITLE
About coordinates space of kp3ds and kp2ds

### DISCUSSION THREAD

**Discussion by 170928** [Arthur151/ROMP#330] [Q&A] [answered]
Hi, @Arthur151 :) 
Thank you for your kind replies. 

I have some questions about data loader! 

In h36m dataset loader, are kp2ds and kp3ds coordinates (in img_info) in World space? 

I can understand kp3ds is Camera Space because it is results of self.smplr or self.map_kps of camkp3d as below:
```python
if self.train_flag and self.regress_smpl:
   verts, kp3ds = self.smplr(pose, beta, gender)
else:
  camkp3d = info['kp3d_mono'].reshape(-1, 3).copy()
  camkp3d -= root_trans
  kp3ds = self.map_kps(camkp3d, maps=self.joint3d_mapper)[None]
```

But, kp2ds is in World Space I think.
```python
kp2d = self.map_kps(info['kp2d'].reshape(-1, 2).copy(), maps=self.joint_mapper) 
kp2ds = np.concatenate([kp2d, self.kps_vis], 1)[None]
```

So, I have a question at this point about coordinate space.
(1) Do kp3ds (Camera Space) and kp2ds (World Space) have different coordinate systems in H36M dataset?
(2) What about the coordinate systems of other datasets? 

And, one more question! 
(3) In h36m, info['cam'] has (3, 3) and info['pose'] has (3, 72).
      What is first dimension of two values ? 
(4) Why you set pose[:3] = root_rotation in h36m. 
      I understand that info['poses'] is axis-angle representation of 24 joints. 
      But, root_rotation is [cam_scale, cam_translation_x_axis, cam_translation_y_axis], not axis-angle representation.

```python
smpl_randidx = 1
root_rotation = np.array(info['cam'])[smpl_randidx]
pose = np.array(info['poses'])[smpl_randidx]
pose[:3] = root_rotation
beta = np.array(info['betas'])
```

**Comment by Arthur151** [comment, marked as answer]
Sorry for the misguidance of the variable names. 

kp3ds is the root-aligned (pelvis) 3D keypoints coordinates.

kp2ds is the 2D keypoints coordinates on the image. 

All datasets follow the same format.

Q: In h36m, info['cam'] has (3, 3) and info['pose'] has (3, 72). What is first dimension of two values.
If the current frame is frame i, 3 is the cam and pose of frame i-1, i, i+1. These numbers are quite similar, which can introduce some noise for supervision, which may avoid overfitting. 

Q: Why you set pose[:3] = root_rotation in h36m.
Here the info['cam'] is the first 3-dim of SMPL pose parameter, global rotation under one of four views of the cameras, instead of the camera parameters we intend to estimate
